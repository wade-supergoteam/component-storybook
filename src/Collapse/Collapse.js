import omit from "lodash.omit";
import PropTypes from "prop-types";
import React, { Component } from "react";
import Transition from "react-transition-group/Transition";
import styled, { css } from "styled-components";
import themeProp from "../utils/theme";
import transitionCollapse from "./default-theme";

const Wrapper = styled.div`
  ${({ status }) => status === "exited" && "display: none;"};
  ${({ status }) => status === "entered" && "display: block;"};

  ${({ status }) =>
    (status === "exiting" || status === "entering") &&
    css`
      height: 0;
      overflow: hidden;
      transition: ${themeProp("transitionCollapse", transitionCollapse)};
    `};
`;

const TransitionTimeouts = {
  Fade: 150, // $transition-fade
  Collapse: 350, // $transition-collapse
  Modal: 300, // $modal-transition
  Carousel: 600, // $carousel-transition
};

// Duplicated Transition.propType keys to ensure that Reactstrap builds
// for distribution properly exclude these keys for nested child HTML attributes
// since `react-transition-group` removes propTypes in production builds.
const TransitionPropTypeKeys = [
  "in",
  "mountOnEnter",
  "unmountOnExit",
  "appear",
  "enter",
  "exit",
  "timeout",
  "onEnter",
  "onEntering",
  "onEntered",
  "onExit",
  "onExiting",
  "onExited",
];

function getHeight(node) {
  return node.scrollHeight;
}

class Collapse extends Component {
  constructor(props) {
    super(props);

    this.state = {
      height: null,
    };
  }

  onEntering = node => {
    this.setState({ height: getHeight(node) });
    this.props.onEntering();
  };

  onEntered = () => {
    this.setState({ height: null });
    this.props.onEntered();
  };

  onExit = node => {
    this.setState({ height: getHeight(node) });
    this.props.onExit();
  };

  onExiting = node => {
    const unused = node.offsetHeight; // eslint-disable-line no-unused-vars
    this.setState({ height: 0 });
    this.props.onExiting();
  };

  onExited = () => {
    this.setState({ height: null });
    this.props.onExited();
  };

  render() {
    const { isOpen, children, ...otherProps } = this.props;
    const { height } = this.state;

    const childProps = omit(otherProps, TransitionPropTypeKeys);

    return (
      <Transition
        {...this.props}
        in={isOpen}
        onEntering={this.onEntering}
        onEntered={this.onEntered}
        onExit={this.onExit}
        onExiting={this.onExiting}
        onExited={this.onExited}
      >
        {status => {
          const style = height === null ? null : { height };
          return (
            <Wrapper status={status} style={{ ...style }} {...childProps}>
              {children}
            </Wrapper>
          );
        }}
      </Transition>
    );
  }
}

Collapse.defaultProps = {
  ...Transition.defaultProps,
  isOpen: false,
  timeout: TransitionTimeouts.Collapse,
};

Collapse.propTypes = {
  ...Transition.propTypes,
  isOpen: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
};

export default Collapse;
