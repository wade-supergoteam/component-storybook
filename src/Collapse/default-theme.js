// @flow

const transitionCollapse = "height 0.35s ease";

export default transitionCollapse;
