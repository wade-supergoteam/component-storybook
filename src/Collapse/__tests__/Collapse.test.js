import React from "react";
import renderer from "react-test-renderer";
import Collapse from "../";

test("it works", () => {
    const wrapper = renderer.create(<Collapse />);
    expect(wrapper).toMatchSnapshot();
});
