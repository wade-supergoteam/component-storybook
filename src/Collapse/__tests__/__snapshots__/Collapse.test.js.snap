// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse with status entered 1`] = `
<div
  className="sc-bdVaJa gtywye"
  style={Object {}}
/>
`;

exports[`Collapse with status exited 1`] = `
<div
  className="sc-bdVaJa bkyCUw"
  style={Object {}}
/>
`;

exports[`Collapse with status exiting 1`] = `
<div
  className="sc-bdVaJa hBMUMD"
  style={Object {}}
/>
`;

exports[`it works 1`] = `
<div
  className="sc-bdVaJa bkyCUw"
  style={Object {}}
/>
`;
